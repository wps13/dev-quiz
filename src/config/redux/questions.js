import questions from '../../utils/questions';
import calculateResult from '../../utils/calculateResult';

const ADD_ANSWER = 'ADD_ANSWER';
const RESET_DATA = 'RESET_DATA';

export const changeQuestion = answerText => ({
  type: ADD_ANSWER,
  payload: {
    answerText,
  },
});

export const reset = () => ({type: RESET_DATA});

const INITIAL_STATE = {
  allQuestions: questions,
  currentQuestion: 0,
  answers: [],
  actualQuestion: questions[0],
  actualQuestionId: 1,
  lastQuestion: false,
  totalQuestions: questions.length,
  correctAnswers: 0,
  minimumPercent: 75,
  correctPercent: 0,
  remainingQuestions: questions.length - 1,
  status: '',
};

const questionReducer = (action, state = INITIAL_STATE) => {
  switch (action.type) {
    case ADD_ANSWER:
      let newAnswers = state.answers.concat({
        id: state.actualQuestionId,
        answer: action.payload.answerText,
      });
      let currentQuestion = state.currentQuestion;
      let totalQ = state.totalQuestions;
      let minimumPerc = state.minimumPercent;
      const ques = state.allQuestions;
      let [corrects, percentCorrect, currentStatus] = calculateResult({
        questions: ques,
        answers: newAnswers,
        totalQuestions: totalQ,
        minimumPercent: minimumPerc,
      });
      let remainingQ = state.remainingQuestions;
      let actualQID = state.actualQuestionId;

      return {
        ...state,
        currentQuestion: currentQuestion + 1,
        answers: newAnswers,
        actualQuestionId: actualQID + 1,
        actualQuestion: questions[currentQuestion + 1],
        correctAnswers: corrects,
        correctPercent: percentCorrect,
        remainingQuestions: remainingQ - 1,
        status: currentStatus,
        lastQuestion: actualQID + 1 >= totalQ,
      };
    case RESET_DATA:
      return {
        ...state,
        answers: [],
        actualQuestion: state.allQuestions[0],
        actualQuestionId: 1,
        lastQuestion: false,
        currentQuestion: 0,
        correctAnswers: 0,
        correctPercent: 0,
        status: '',
        remainingQuestions: state.allQuestions.length - 1,
      };
    default:
      return state;
  }
};

export default questionReducer;
