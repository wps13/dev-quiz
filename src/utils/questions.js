export default [
  {
    id: 1,
    title: 'Which tag is used to insert an image in html?',
    alternatives: ['p', 'img', 'div', 'media'],
    answer: 'img',
  },
  {
    id: 2,
    title: 'How does position absolute works on css?',
    alternatives: [
      'It moves the element out of the document flow',
      'it moves the elements to another parent',
      'it hides the elements',
      'it puts the element above all the other elements',
    ],
    answer: 'It moves the element out of the document flow',
  },
  {
    id: 3,
    title: 'How to grow the flex container?',
    alternatives: ['flex-grow', 'flex-flow', 'flex', 'flex-wrap'],
    answer: 'flex-grow',
  },
];
