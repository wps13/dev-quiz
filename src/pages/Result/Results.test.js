import React from 'react';
import {shallow, mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';

import Results from './Result';

import questions from '../../../__mocks__/questions';

const mockStore = configureMockStore([]);

const defaultValue = {
  allQuestions: questions,
  currentQuestion: 1,
  answers: ['img'],
  actualQuestion: questions[1],
  actualQuestionId: 2,
  lastQuestion: true,
  totalQuestions: questions.length,
  correctAnswers: 1,
  minimumPercent: 75,
  correctPercent: 50,
  remainingQuestions: 0,
  status: 'failed',
};

const store = mockStore({question: defaultValue});

store.dispatch = jest.fn();

const props = {
  navigation: {
    navigate: jest.fn(),
  },
};

describe('Results', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <Results {...props} />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render the texts correctly', () => {
    jest.mock('react-redux', () => ({
      useSelector: selectorFn =>
        selectorFn({
          question: {
            allQuestions: [
              {
                id: 1,
                title: 'Which tag is used to insert an image in html?',
                alternatives: ['p', 'img', 'div', 'media'],
                answer: 'img',
              },
              {
                id: 2,
                title: 'How does position absolute works on css?',
                alternatives: [
                  'It moves the element out of the document flow',
                  'it moves the elements to another parent',
                  'it hides the elements',
                  'it puts the element above all the other elements',
                ],
                answer: 'It moves the element out of the document flow',
              },
            ],
            currentQuestion: 1,
            answers: ['img'],
            actualQuestion: {
              id: 2,
              title: 'How does position absolute works on css?',
              alternatives: [
                'It moves the element out of the document flow',
                'it moves the elements to another parent',
                'it hides the elements',
                'it puts the element above all the other elements',
              ],
              answer: 'It moves the element out of the document flow',
            },
            actualQuestionId: 2,
            lastQuestion: true,
            totalQuestions: 2,
            correctAnswers: 1,
            minimumPercent: 75,
            correctPercent: 50,
            remainingQuestions: 0,
            status: 'failed',
          },
        }),
      useDispatch: jest.fn(),
    }));
    const wrapper = mount(
      <Provider store={store}>
        <Results {...props} />
      </Provider>,
    );
    const title = wrapper.findWhere(
      node => node.prop('testID') === 'Results-title',
    );

    const description = wrapper.findWhere(
      node => node.prop('testID') === 'Results-description',
    );
    const correctPercent = wrapper.findWhere(
      node => node.prop('testID') === 'Results-correctPercent',
    );

    const minimumPercent = wrapper.findWhere(
      node => node.prop('testID') === 'Results-minimumPercent',
    );

    expect(title.exists()).toEqual(true);
    expect(title.first().text()).toEqual('You failed!');

    expect(description.exists()).toEqual(true);
    expect(description.first().text()).toEqual(
      'You answered 1 out of 2 correctly',
    );

    expect(correctPercent.exists()).toEqual(true);
    expect(correctPercent.first().text()).toEqual('You got 50% correct');

    expect(minimumPercent.exists()).toEqual(true);
    expect(minimumPercent.first().text()).toEqual(
      'Minimum required to pass: 75%',
    );
  });
  it('should call go to start correctly', () => {
    jest.mock('react-redux', () => ({
      useSelector: selectorFn =>
        selectorFn({
          question: {
            allQuestions: [
              {
                id: 1,
                title: 'Which tag is used to insert an image in html?',
                alternatives: ['p', 'img', 'div', 'media'],
                answer: 'img',
              },
              {
                id: 2,
                title: 'How does position absolute works on css?',
                alternatives: [
                  'It moves the element out of the document flow',
                  'it moves the elements to another parent',
                  'it hides the elements',
                  'it puts the element above all the other elements',
                ],
                answer: 'It moves the element out of the document flow',
              },
            ],
            currentQuestion: 1,
            answers: ['img'],
            actualQuestion: {
              id: 2,
              title: 'How does position absolute works on css?',
              alternatives: [
                'It moves the element out of the document flow',
                'it moves the elements to another parent',
                'it hides the elements',
                'it puts the element above all the other elements',
              ],
              answer: 'It moves the element out of the document flow',
            },
            actualQuestionId: 2,
            lastQuestion: true,
            totalQuestions: 2,
            correctAnswers: 1,
            minimumPercent: 75,
            correctPercent: 50,
            remainingQuestions: 0,
            status: 'failed',
          },
        }),
      useDispatch: jest.fn(),
    }));
    const wrapper = mount(
      <Provider store={store}>
        <Results {...props} />
      </Provider>,
    );
    const button = wrapper.findWhere(
      node => node.prop('testID') === 'Results-buttonHome',
    );

    button.props().pressFunction();

    expect(button.exists()).toEqual(true);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
