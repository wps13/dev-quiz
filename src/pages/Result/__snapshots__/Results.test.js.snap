// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Result
    navigation={
      Object {
        "navigate": [MockFunction],
      }
    }
  />
</ContextProvider>
`;
